pkg/controllers/plugin/pluginpreset_controller.go

```diff
@@ -62,79 +63,84 @@ func (r *PluginPresetReconciler) SetupWithManager(name string, mgr ctrl.Manager)
 	default:
-		pluginPresetStatus.SetConditions(greenhousev1alpha1.FalseCondition(greenhousev1alpha1.ClusterListEmpty, "", ""))
+		pluginPreset.SetCondition(greenhousev1alpha1.FalseCondition(greenhousev1alpha1.ClusterListEmpty, "", ""))
+	}
+
+	skippedCondition, failedCondition, err := r.reconcilePluginPreset(ctx, pluginPreset, clusters)
+	pluginPreset.SetCondition(skippedCondition)
+	pluginPreset.SetCondition(failedCondition)
+	if err != nil {
+		return ctrl.Result{}, lifecycle.Failed, err
 	}
 
-	skippedCondition, failedCondition, err := r.reconcilePluginPreset(ctx, pluginPreset, clusters, pluginPresetStatus)
-	pluginPresetStatus.SetConditions(skippedCondition, failedCondition)
+	return ctrl.Result{}, lifecycle.Success, nil
+}
+
+func (r *PluginPresetReconciler) EnsureDeleted(ctx context.Context, resource lifecycle.RuntimeObject) (ctrl.Result, lifecycle.ReconcileResult, error) {
+	pluginPreset := resource.(*greenhousev1alpha1.PluginPreset) //nolint:errcheck
+
+	// Cleanup the plugins that are managed by this PluginPreset.
+	plugins, err := r.listPlugins(ctx, pluginPreset)
 	if err != nil {
-		return ctrl.Result{}, err
+		return ctrl.Result{}, lifecycle.Failed, err
+	}
+	allErrs := make([]error, 0)
+	for _, plugin := range plugins.Items {
+		if err := r.Client.Delete(ctx, &plugin); err != nil && !apierrors.IsNotFound(err) {
+			allErrs = append(allErrs, err)
+		}
+	}
+
+	// If there are still plugins left, requeue the deletion.
+	if len(allErrs) > 0 {
+		return ctrl.Result{}, lifecycle.Pending, fmt.Errorf("failed to delete plugins for %s/%s: %w", pluginPreset.Namespace, pluginPreset.Name, errors.Join(allErrs...))
 	}
 
-	return ctrl.Result{}, nil
+	return ctrl.Result{}, lifecycle.Success, nil
 }
 
 // reconcilePluginPreset reconciles the PluginPreset by creating or updating the Plugins for the given clusters.
 // It skips reconciliation for Plugins that do not have the labels of the PluginPreset.
-func (r *PluginPresetReconciler) reconcilePluginPreset(ctx context.Context, preset *greenhousev1alpha1.PluginPreset, clusters *greenhousev1alpha1.ClusterList, status greenhousev1alpha1.PluginPresetStatus) (skippedCondition, failedCondition greenhousev1alpha1.Condition, err error) {
+func (r *PluginPresetReconciler) reconcilePluginPreset(ctx context.Context, preset *greenhousev1alpha1.PluginPreset, clusters *greenhousev1alpha1.ClusterList) (skippedCondition, failedCondition greenhousev1alpha1.Condition, err error) {
 	var allErrs = make([]error, 0)
 	var skippedPlugins = make([]string, 0)
 	var failedPlugins = make([]string, 0)
 
-	failedCondition = *status.GetConditionByType(greenhousev1alpha1.PluginFailedCondition)
-	skippedCondition = *status.GetConditionByType(greenhousev1alpha1.PluginSkippedCondition)
+	failedCondition = *preset.Status.GetConditionByType(greenhousev1alpha1.PluginFailedCondition)
+	skippedCondition = *preset.Status.GetConditionByType(greenhousev1alpha1.PluginSkippedCondition)
```

Comment:
Since we are now able to set the status directly on the resource being reconciled, it is no longer necessary to return the status conditions. You can remove the conditions from the return of the function.

Instead of getting and updating the conditions we can directly set them on the PluginPreset here: https://github.com/cloudoperators/greenhouse/pull/722/files?diff=split&w=0#diff-4936a1860e5afa112e0714d466d89465f57178da372b0c301a4f9ccc81675620R195-R210